Scenes: Annotations functionality in settings

mdvictor opened the pull request:
This PR implements CRUD operations for annotations in scenes dashboard settings. Part of https://github.com/grafana/grafana/issues/80891

It migrates addition,deletion,moving and editing of annotations.

It needs [these](https://github.com/grafana/scenes/pull/554) modifications in the scenes library to work.

mdvictor committed:
wip listing/viewing annotations

mdvictor committed:
list annotations in settings

mdvictor committed:
fix tests

mdvictor committed:
wip edit mode

mdvictor committed:
PR mods

mdvictor committed:
Merge branch 'mdvictor/scenes-annotation-settings-list' into mdvictor/scenes-annotations-settings-edit

mdvictor committed:
move, delete, edit, add new

mdvictor committed:
edit annotation in settings

mdvictor committed:
Annotations functionality

mdvictor committed:
Merge branch 'main' into mdvictor/scenes-annotations-settings-edit

mdvictor committed:
revert change

mdvictor committed:
add ui tests, move angularEditorLoader

mdvictor committed:
Merge branch 'main' into mdvictor/scenes-annotations-settings-edit

mdvictor committed:
remove flaky test

mdvictor commented on public/app/features/dashboard-scene/settings/AnnotationsEditView.tsx:
```diff
@@ -36,58 +37,136 @@ export class AnnotationsEditView extends SceneObjectBase<AnnotationsEditViewStat
+
+    const data = this.getSceneDataLayers();
+
+    if (data) {
```
This actually needs to be changed since we're now throwing an error if there are no scene data layers

mdvictor committed:
refactor

dprokop commented on public/app/features/dashboard-scene/settings/AnnotationsEditView.tsx:
```diff
@@ -36,58 +37,130 @@ export class AnnotationsEditView extends SceneObjectBase<AnnotationsEditViewStat
+      //see the annotation on the panel
+      layer.runLayer();
+
+      data.setState({
+        layers,
+      });
```
Is this needed? Since we are updating the layer the subscription to it should already be in place in the data provider, so any updates after `runLayer` should be picked up by this subscription.

dprokop commented on public/app/features/dashboard-scene/scene/DashboardScene.tsx:
```diff
@@ -329,6 +330,22 @@ export class DashboardScene extends SceneObjectBase<DashboardSceneState> {
     return this.state.viewPanelScene ?? this.state.body;
   }
 
+  public getVizPanels(): VizPanel[] {
+    const panels: VizPanel[] = [];
+
+    this.state.body.forEachChild((child) => {
+      if (child instanceof SceneGridItem) {
+        const panel = child.state.body;
+
+        if (panel instanceof VizPanel) {
+          panels.push(panel);
+        }
+      }
+    });
+
+    return panels;
+  }
```
This implementation isn't fully correct. SceneGridItems can also be part a child of SceneGridRow. Please have a look in the `transformSaveModelToScene` how the rows are created.

Also - can you please move this function from DashboardScene to `dashboardSceneGraph` util?

dprokop reviewed (commented):
Code wise this looks good, however i found two bugs:
1. Say you are toggling a layer on in your dashboard. Then enter the edit mode and Discard. The toggle doesn't work anymore (unless the dashboard is refreshed). See the video:

https://github.com/grafana/grafana/assets/2376619/bdbf39f5-009c-4167-992b-55f8b20a5ac9

2. When going to a particular annotation editor, pressing back button does not navigate to the annotations list, but to general settings instead. I think we have this problem also visible in variables and links edit. I'll  create a bug report we can tackle indpendently form this PR.

ivanortegaalba commented on public/app/features/dashboard-scene/settings/AnnotationsEditView.tsx:
```diff
@@ -36,58 +37,130 @@ export class AnnotationsEditView extends SceneObjectBase<AnnotationsEditViewStat
+      iconColor: 'red',
+    };
+
+    const newAnnotation = new DashboardAnnotationsDataLayer({
+      key: `annotations-${newAnnotationQuery.name}`,
+      query: newAnnotationQuery,
+      name: newAnnotationQuery.name,
+      isEnabled: Boolean(newAnnotationQuery.enable),
+      isHidden: Boolean(newAnnotationQuery.hide),
+    });
+
+    const data = this.getSceneDataLayers();
+
+    const layers = [...data.state.layers];
+    newAnnotation.activate();
+
+    //keep annotation layers together
+    layers.splice(this.getAnnotationsLength(), 0, newAnnotation);
+
+    data.setState({
+      layers,
+    });
+
+    this.setState({ editIndex: this.getAnnotationsLength() - 1 });
```
I don't fully understand why we need this, why don't we do something like this:
```suggestion
    const newAnnotation = new DashboardAnnotationsDataLayer({
      key: `annotations-${newAnnotationQuery.name}`,
      query: newAnnotationQuery,
      name: newAnnotationQuery.name,
      isEnabled: Boolean(newAnnotationQuery.enable),
      isHidden: Boolean(newAnnotationQuery.hide),
    });
    newAnnotation.activate();

    const data = this.getSceneDataLayers();

    const layers = [...data.state.layers, newAnnotation];

    data.setState({
      layers,
    });

    this.setState({ editIndex: this.getAnnotationsLength() - 1 });
```

Not fully sure, but shouldn't we activate it once the object is in the hierarchy? just in case the activation handler needs anything from the parent.

mdvictor replied:
I'm not sure I follow, are you suggesting activating the new annotation after it's set in the layers? Also the splice is there to keep annotations grouped since there might be other data layer types in between.

ivanortegaalba replied:
Sorry Victor, let me explain 🙇 

> I'm not sure I follow, are you suggesting activating the new annotation after it's set in the layers?

For the scene objects, we usually have activation handlers. They may access the parents or the dashboard scene. Activating the scene object not being part of the dashboard scene yet could cause an error as the dashboard scene won't be found. Probably, in this case, the handler doesn't need anything from the parent, but I'd consider it as a good practice.

> Also the splice is there to keep annotations grouped since there might be other data layer types in between.

When you say grouped, what do you mean? I don't understand why it is different from creating a new array and appending the new annotation in the end or even using `push()`. 
AFAIK, if I have `a = [annot1, annot2]` and I do `[...a, annot3]`, the output is `[annot1, annot2, annot3]` same as using `splice()` or `push()`. 
I mean, it works, just I'm curious about it since it is not a very common way to add new methods when you create a new array anyway 😄

ivanortegaalba commented:
In terms of annotation state, it works like a charm. I just found a couple of things missing:

1. The breadcrumb is not updated when entering into an annotation
![Captura de pantalla 2024-02-02 a las 12 03 28](https://github.com/grafana/grafana/assets/5699976/0839e9f7-cbeb-4938-bacd-f82d9fde4ef0)

2. DashboardScene does not detect the changes. When changing the annotations state the "save" button should be enabled. This can go in a different PR to not make this PR bigger. [This function](https://github.com/grafana/grafana/blob/main/public/app/features/dashboard-scene/scene/DashboardScene.tsx#L332) detects changes from every scene object and sets isDirty to true when the change involves the dashboard updating.

mdvictor committed:
bump scenes version, refactor getVizPanels, refactor edit page

mdvictor committed:
Merge branch 'main' into mdvictor/scenes-annotations-settings-edit

mdvictor committed:
fix nav breadcrumbs

mdvictor commented:
Thank you for all the great insights from your review!

@dprokop, the enter/exit edit mode toggle issue is now fixed. The PR wasn't using the latest scenes lib. 

@ivanortegaalba breadcrumbs are now fixed, still need to look into turning on the isDirty flag for annotations and fix that either here or in a new PR

mdvictor committed:
annotation set dirty flag, add overlay to edit view

mdvictor commented:
Added tracking for annotations and the save drawer overlay in annotations settings

mdvictor committed:
Merge branch 'main' into mdvictor/scenes-annotations-settings-edit

ivanortegaalba commented on public/app/features/dashboard-scene/settings/AnnotationsEditView.tsx:
```diff
@@ -36,72 +46,221 @@ export class AnnotationsEditView extends SceneObjectBase<AnnotationsEditViewStat
+        editIndex={editIndex}
+        navModel={navModel}
+        dashboard={dashboard}
+        overlay={overlay}
```
Torkel opened this PR #81852 which moves the overlay out of the edit views, so we won't need to add it in the edit views anymore 🙌

mdvictor replied:
Removed the overlays

ivanortegaalba commented on public/app/features/dashboard-scene/scene/DashboardScene.tsx:
```diff
@@ -338,6 +340,14 @@ export class DashboardScene extends SceneObjectBase<DashboardSceneState> {
             this.setIsDirty();
           }
         }
+        if (event.payload.changedObject instanceof SceneDataLayers) {
+          this.setIsDirty();
+        }
+        if (event.payload.changedObject instanceof dataLayers.AnnotationsDataLayer) {
+          if (Object.prototype.hasOwnProperty.call(event.payload.partialUpdate, 'query')) {
+            this.setIsDirty();
+          }
+        }
```
I just opened a PR changing the approach of detecting changes #81958, I'll add this there since the approach of detecting changes is different to ensure the changes are really made. 

There is a difference there, and it is I only check if it is an instance of `dataLayers.AnnotationsDataLayer` to be able to save if it is enabled or not as the default value.

Do you think it makes sense?

mdvictor replied:
@ivanortegaalba I think we need the `query` property verification in `partialUpdate` because the datalayer will update the `data` property with `PanelData` that might be detected as a dashboard change, when it really isn't. We only care to set the scene dirty when actual annotation query modifications appear in annotations edit page, and these modifications we are interested in live in the `query` property

ivanortegaalba replied:
There is one use case not covered here. When we enable or disable the annotation, from the toolbar or the settings, it won't be detected as dirty state, so we won't be able to save that as default values for the dashboard, as we do for variable values.

I'd suggest to do something like 
```js
if (event.payload.changedObject instanceof dataLayers.AnnotationsDataLayer) {
          if (!Object.prototype.hasOwnProperty.call(event.payload.partialUpdate, '$data')) {
            this.setIsDirty();
          }
        }
```
I don't know if `$data` is the right prop to contain, but we can just black list the properties that change when the values are resolved. WDYT?

ivanortegaalba commented on public/app/features/dashboard-scene/settings/AnnotationsEditView.tsx:
```diff
@@ -1,24 +1,34 @@
+import { AnnotationSettingsEdit, AnnotationSettingsList, newAnnotationName } from './annotations';
 import { DashboardEditView, DashboardEditViewState, useDashboardEditPageNav } from './utils';
 
+export enum MoveDirection {
```
I like this 💯

ivanortegaalba commented on public/app/features/dashboard-scene/settings/AnnotationsEditView.tsx:
```diff
@@ -36,72 +46,221 @@ export class AnnotationsEditView extends SceneObjectBase<AnnotationsEditViewStat
+  };
+
+  public onUpdate = (annotation: AnnotationQuery, editIndex: number) => {
+    const data = this.getSceneDataLayers();
+
+    const layers = [...data.state.layers];
+    const layer = layers[editIndex];
```
```suggestion
    const layer = this.getDataLayer(editIndex);
```
Maybe we can use the layer-getter here, WDYT?

ivanortegaalba commented on public/app/features/dashboard-scene/settings/annotations/AnnotationSettingsEdit.test.tsx:
```diff
@@ -0,0 +1,278 @@
+      renderer: { getByTestId },
+    } = setup();
+
+    await waitFor(() => {
```
If you want to avoid all of `waitFor` you can always wrap the `render()` with `act()`, so you ensure the react component went through all the rendering cycles and it already ran all the async functions 😄 See [here](https://github.com/grafana/grafana/pull/81318/files#diff-cd58a512daf25cfb1e36bc87b1e667e0421f69aa88487a4255c5110ac8058935R72).

Just a tip, you don't need to change everything 🙏

mdvictor replied:
I'll look into this! Thank you!

mdvictor replied:
Refactored and remove `waitFor` calls! 💯

ivanortegaalba commented on public/app/features/dashboard-scene/settings/annotations/AnnotationSettingsEdit.tsx:
```diff
@@ -0,0 +1,339 @@
+            Delete
+          </Button>
+        )}
+        <Button
+          variant="secondary"
+          onClick={onPreview}
+          data-testid={selectors.pages.Dashboard.Settings.Annotations.NewAnnotation.previewInDashboard}
+        >
+          Preview in dashboard
+        </Button>
+        <Button
+          variant="primary"
+          onClick={onApply}
+          data-testid={selectors.pages.Dashboard.Settings.Annotations.NewAnnotation.apply}
+        >
+          Apply
+        </Button>
```
We can replace both buttons with "Back to list", since the new nav bar contains `Back to dashboard`, the same as the preview dashboard. Also, we agreed with @Ijin08 to not have apply buttons and have back-to-list ones.

As we do for variables (you will see links keeps the apply, I'm opening a PR to fix it)
<img width="1712" alt="Captura de pantalla 2024-02-07 a las 8 56 00" src="https://github.com/grafana/grafana/assets/5699976/39738312-3328-4f13-b392-9f9f97942dbb">

```suggestion
        <Button
          onClick={onBackToList}
          data-testid={selectors.pages.Dashboard.Settings.Annotations.NewAnnotation.apply}
        >
          Back to list
        </Button>
```

ivanortegaalba reviewed (commented):
I left some comments, due to some changes that were recently made. Amazing work @mdvictor with all of this! Testing locally! 🙇

mdvictor commented:
@ivanortegaalba thank you the explanations around layer activation. I will modify and activate the layer after setting the layers state. 

As for the splice there, let me explain: 
From what I understand there are multiple types of data layers that live in that list. IIRC there are annotations, alerts and transformations related data layers. On a new dashboard, the list is populated with a default annotation and a default alert, so the list has those two in there in this order.

When creating a new annotation I use the splice to add that annotation right after the last annotation layer so the list results in something like this: [defaultAnno, someOtherAnno, newAnno, defaultAlert, someOtherAlert].

Just pushing the annotation at the end of the list will cause issues with the `editIndex`, since an anno in the annotation list settings page might appear at position 3, for example, but in the datalayers list it's actually on a different position because other type of layers are in between.

mdvictor committed:
PR mods

dprokop commented on public/app/features/dashboard-scene/settings/AnnotationsEditView.tsx:
```diff
@@ -36,72 +45,210 @@ export class AnnotationsEditView extends SceneObjectBase<AnnotationsEditViewStat
-  }
-
-  public getSceneDataLayers(): SceneDataLayers | undefined {
+  public getSceneDataLayers(): SceneDataLayers {
```
Can we move this util to `dashboardSceneGraph.getDataLayers`? again - to handle the potential structural changes in the future.

mdvictor committed:
PR mods

mdvictor committed:
remove flaky test

mdvictor committed:
change dirty flag setting logic for anotations

ivanortegaalba commented on public/app/features/dashboard-scene/settings/annotations/AnnotationSettingsEdit.tsx:
```diff
@@ -0,0 +1,317 @@
+            Delete
+          </Button>
+        )}
+        <Button onClick={onBackToList} data-testid={selectors.pages.Dashboard.Settings.Annotations.NewAnnotation.apply}>
```
```suggestion
        <Button variant="secondary" onClick={onBackToList} data-testid={selectors.pages.Dashboard.Settings.Annotations.NewAnnotation.apply}>
```
![Captura de pantalla 2024-02-12 a las 12 54 16](https://github.com/grafana/grafana/assets/5699976/aba0f1a5-e7c3-43fd-8252-0cba02a0854d)
By default, the variant is primary, but for back to list we're using secondary :)

ivanortegaalba reviewed (approved):
@mdvictor I tested all functionality manually and it works perfectly fine!

I used `gdev dashboards > Annotation filtering` to test the following:
✅ Update name, hide, enable properties for annotations
✅ Filter where to display annotations
✅ Map props to annotations attributes
✅ CRUD actions on list

I just suggested a couple of design changes to adjust to the rest of the settings views

ivanortegaalba commented on public/app/features/dashboard-scene/settings/annotations/AnnotationSettingsList.tsx:
```diff
@@ -2,19 +2,24 @@ import { css } from '@emotion/css';
   onDelete: (idx: number) => void;
 };
 
+export const BUTTON_TITLE = 'Add annotation query';
```
[`<ListNewButton>`](https://github.com/grafana/grafana/pull/81361/files#diff-9de303fbd0fd464b628e673a0370bb0aea8c4889ec39feba48c6394d7e052ec5L111) should be moved to `primary`.

I just realized the empty states' copies use `Add` but the buttons use `New`.

We should make them consistent across views. I'll ask Isabel about what to do there, let's not block this PR for that inconsistency, I can solve it in a second PR.

mdvictor replied:
Yes, I was just about to mention that the CTA buttons across the settings refer to `Add`, not `New`. I will leave this one as it was, just for consistency, and we can come back and refactor all pages later.

I will, however, update the `ListNewButton` to be primary

mdvictor replied:
ListNewButton modification was already merged, thank you! :D

mdvictor committed:
change button variants

mdvictor committed:
Merge branch 'main' into mdvictor/scenes-annotations-settings-edit
